### ISSUE TITLE
how to calulate the pos_u?

### ISSUE THREAD

**Issue by Ouz-2** [Jinfeng-Xu/NLGCL#1]
def neighbor_cl_loss(self, emb_list, users, pos_items, neg_items):
        """Compute neighbor layer contrastive loss"""
        # Get initial embeddings (layer 0)
        user_emb_0, item_emb_0 = torch.split(emb_list[0], [self.n_users, self.n_items])
        cl_user_loss = 0.0
        cl_item_loss = 0.0

        # Compute contrastive loss at each layer
        for layer_idx in range(1, self.n_layers + 1):
            # Get current layer embeddings
            user_emb_k, item_emb_k = torch.split(emb_list[layer_idx], [self.n_users, self.n_items])
            
            # User-side contrastive loss
            cl_user_loss = self.InfoNCE(
                anchor=item_emb_k[pos_items],
                positive=user_emb_0[users],
                all_samples=user_emb_0[users]
            ) + 1e-6
           
            cl_user_loss += pos_u - neg_u + 1e-6  # Small constant for stability
            
            # Item-side contrastive loss
            cl_item_loss = self.InfoNCE(
                anchor=user_emb_k[users],
                positive=item_emb_0[pos_items],
                all_samples=item_emb_0[pos_items]
            ) + 1e-6
            
            # Update reference embeddings for next layer
            user_emb_0, item_emb_0 = user_emb_k, item_emb_k

this work is very helpful.but i want to know how to calculate the pos_u and neg_u in the " cl_user_loss += pos_u - neg_u + 1e-6  "?

**Comment by Jinfeng-Xu** [comment]
> def neighbor_cl_loss(self, emb_list, users, pos_items, neg_items): """Compute neighbor layer contrastive loss""" # Get initial embeddings (layer 0) user_emb_0, item_emb_0 = torch.split(emb_list[0], [self.n_users, self.n_items]) cl_user_loss = 0.0 cl_item_loss = 0.0
> 
> ```
>     # Compute contrastive loss at each layer
>     for layer_idx in range(1, self.n_layers + 1):
>         # Get current layer embeddings
>         user_emb_k, item_emb_k = torch.split(emb_list[layer_idx], [self.n_users, self.n_items])
>         
>         # User-side contrastive loss
>         cl_user_loss = self.InfoNCE(
>             anchor=item_emb_k[pos_items],
>             positive=user_emb_0[users],
>             all_samples=user_emb_0[users]
>         ) + 1e-6
>        
>         cl_user_loss += pos_u - neg_u + 1e-6  # Small constant for stability
>         
>         # Item-side contrastive loss
>         cl_item_loss = self.InfoNCE(
>             anchor=user_emb_k[users],
>             positive=item_emb_0[pos_items],
>             all_samples=item_emb_0[pos_items]
>         ) + 1e-6
>         
>         # Update reference embeddings for next layer
>         user_emb_0, item_emb_0 = user_emb_k, item_emb_k
> ```
> 
> this work is very helpful.but i want to know how to calculate the pos_u and neg_u in the " cl_user_loss += pos_u - neg_u + 1e-6 "?

Thank you for your attention. You can ignore this line; we will update the repository to remove it. This was part of our follow-up exploration. It was mistakenly left in the code while restructuring it to improve readability after NLGCL was accepted.
